Hello there!

I going to use Market Basket Analysis using Association Rule-Mining on Groceries Data which is available in [here](https://github.com/stedy/Machine-Learning-with-R-datasets/blob/master/groceries.csv). 

Before jumping to coding, a quick question have you ever wonder how you end up with purchasing items/things which were not in your shopping list during shopping at market/Dmart? It happens due to the chances of buying products are related. When you bought a loaf of bread, you saw butter on the next step. These items are not same but the relationship between them tend to increase the chance of buying them. Through association rule mining we are going to determine the strength of relation between two or more different object using Apriori Algorithm. 

**Applications of Association Rule Mining **
It has various applications in machine learning and data mining. Below are some popular applications of association rule learning:

*Market Basket Analysis:* It is one of the popular examples and applications of association rule mining. This technique is commonly used by big retailers to determine the association between items.

*Medical Diagnosis:* With the help of association rules, patients can be cured easily, as it helps in identifying the probability of illness for a particular disease.

*Protein Sequence:* The association rules help in determining the synthesis of artificial Proteins.

It is also used for the *Catalog Design* and *Loss-leader Analysis* and many more other applications.

**Apriori Algorithm**
This algorithm was given by the R. Agrawal and Srikant in the year 1994. It is mainly used for market basket analysis and helps to find those products that can be bought together. It can also be used in the healthcare field to find drug reactions for patients.

The apriori algorithm finds frequent itemsets that occur atleast as frequently as minimum support count to generate association rules which will prune rules which fail minimum support and minimum confidence threshold.

Before we proceed with market basket analysis there are some terms which we need to know.

**Support:** Support is an indication of how frequently the item/itemset appears in a transaction/dataset.

**Confidence:** If we were given two products X and Y such that for rule X(antecedents)=>Y(consequents), confidence shows the percentage of  product Y is bought with product X. It's an indication how often the rule is proven correct.

**Lift:** It's a ratio which indicates whether the given two products X & Y are dependent on each other or not. A lift value greater than 1 means that Y is likely to be bought if  X is bought, while a value less than 1 means that Y is unlikely to be bought if X is bought. Lift value 1 indicates no association between products. 

**Problem Statement**
1. Read the CSV file into a data frame. Get the shape.
2. Find the top 20 “sold items” that occur in the dataset.
3. Find how much of the total sales they account for.
4. Create a function prune_dataset, which will help us reduce the size of our dataset based on our requirements. 
5. Implement Apriori Algorithm and create rules.

In [ ]:
import pandas as pd
import numpy as np

from sklearn.preprocessing import LabelEncoder
from collections import Counter

from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
#1. read csv file into data frame. Get the shape.
data = pd.read_csv('https://raw.githubusercontent.com/stedy/Machine-Learning-with-R-datasets/master/groceries.csv',sep='\n',header=None)
data.head(2)

0
0  citrus fruit,semi-finished bread,margarine,rea...
1                       tropical fruit,yogurt,coffee

In [62]:
data.shape

(9835, 1)

In [ ]:
data.values

array([['citrus fruit,semi-finished bread,margarine,ready soups'],
       ['tropical fruit,yogurt,coffee'],
       ['whole milk'],
       ...,
       ['chicken,citrus fruit,other vegetables,butter,yogurt,frozen dessert,domestic eggs,rolls/buns,rum,cling film/bags'],
       ['semi-finished bread,bottled water,soda,bottled beer'],
       ['chicken,tropical fruit,other vegetables,vinegar,shopping bags']],
      dtype=object)

In [ ]:
#top 20 "sold items" that occur in the dataset. total sale they account for
transactions = []
for T in data.values:
    transactions.append(T.tolist())

In [ ]:
itemset = [itemset[0].split(",") for itemset in transactions]
item_name = [item for T in itemset for item in T]
items = Counter(item_name).keys(); itemcounts = Counter(item_name).values()

In [ ]:
df1 = pd.DataFrame(items, columns =['item_name'])
df2 = pd.DataFrame(itemcounts, columns=['item_count'])
goc_df = pd.concat([df1,df2],axis=1)

In [ ]:
goc_df['item_perc'] = goc_df.item_count/goc_df.item_count.sum()
goc_df.sort_values(by='item_count', ascending=False, inplace=True)

In [64]:
goc_df.head(20)

item_name  item_count  item_perc  total_perc
7              whole milk        2513   0.057947    0.057947
11       other vegetables        1903   0.043881    0.101829
17             rolls/buns        1809   0.041714    0.143542
31                   soda        1715   0.039546    0.183089
5                  yogurt        1372   0.031637    0.214725
24          bottled water        1087   0.025065    0.239791
42        root vegetables        1072   0.024719    0.264510
4          tropical fruit        1032   0.023797    0.288307
52          shopping bags         969   0.022344    0.310651
50                sausage         924   0.021307    0.331957
39                 pastry         875   0.020177    0.352134
0            citrus fruit         814   0.018770    0.370904
19           bottled beer         792   0.018263    0.389167
35             newspapers         785   0.018101    0.407268
49            canned beer         764   0.017617    0.424885
8               pip fruit         744   0.017156    0.442041
34  fruit/vegetable juice         711   0.016395    0.458436
60     whipped/sour cream         705   0.016257    0.474693
51            brown bread         638   0.014712    0.489404
65          domestic eggs         624   0.014389    0.503793

In [ ]:
goc_df['total_perc'] = goc_df['item_perc'].cumsum()
top20_trans = goc_df.head(20)
top20_trans.reset_index(drop=True)

item_name  item_count  item_perc  total_perc
0              whole milk        2513   0.057947    0.057947
1        other vegetables        1903   0.043881    0.101829
2              rolls/buns        1809   0.041714    0.143542
3                    soda        1715   0.039546    0.183089
4                  yogurt        1372   0.031637    0.214725
5           bottled water        1087   0.025065    0.239791
6         root vegetables        1072   0.024719    0.264510
7          tropical fruit        1032   0.023797    0.288307
8           shopping bags         969   0.022344    0.310651
9                 sausage         924   0.021307    0.331957
10                 pastry         875   0.020177    0.352134
11           citrus fruit         814   0.018770    0.370904
12           bottled beer         792   0.018263    0.389167
13             newspapers         785   0.018101    0.407268
14            canned beer         764   0.017617    0.424885
15              pip fruit         744   0.017156    0.442041
16  fruit/vegetable juice         711   0.016395    0.458436
17     whipped/sour cream         705   0.016257    0.474693
18            brown bread         638   0.014712    0.489404
19          domestic eggs         624   0.014389    0.503793

In [ ]:
#transforming item_name rows to columns.
TE = TransactionEncoder()
tran_data = TE.fit_transform(itemset)
tran_df = pd.DataFrame(tran_data, columns=TE.columns_)
tran_df

Instant food products  UHT-milk  ...  yogurt  zwieback
0                     False     False  ...   False     False
1                     False     False  ...    True     False
2                     False     False  ...   False     False
3                     False     False  ...    True     False
4                     False     False  ...   False     False
...                     ...       ...  ...     ...       ...
9830                  False     False  ...   False     False
9831                  False     False  ...   False     False
9832                  False     False  ...    True     False
9833                  False     False  ...   False     False
9834                  False     False  ...   False     False

[9835 rows x 169 columns]

In [ ]:
#Label Encoding.
columns = tran_df.columns
gle = LabelEncoder()
for i in tran_df.columns:
  tran_df[i] = gle.fit_transform(tran_df[i])
tran_df.head(2)

Instant food products  UHT-milk  ...  yogurt  zwieback
0                      0         0  ...       0         0
1                      0         0  ...       1         0

[2 rows x 169 columns]

In [ ]:
#Data pruning
def prune_dataset(input_df, length_trans, sales_perc):
    transactions = []
    for T in input_df.values:
        transactions.append(T.tolist())
    itemset = [itemset[0].split(",") for itemset in transactions]
    item_name = [item for T in itemset for item in T]
    items = Counter(item_name).keys(); itemcounts = Counter(item_name).values()
    
    data1 = pd.DataFrame(items, columns =['item_name'])
    data2 = pd.DataFrame(itemcounts, columns=['item_count'])
    new_df = pd.concat([data1,data2],axis=1)
    
    new_df['item_perc'] = new_df.item_count/new_df.item_count.sum()
    new_df.sort_values(by='item_count', ascending=False, inplace=True)
    new_df['total_perc'] = new_df['item_perc'].cumsum()
    
    top_items = list(new_df[new_df.total_perc < sales_perc].item_name)

    TE = TransactionEncoder()
    tran_data = TE.fit_transform(itemset)
    tran_df = pd.DataFrame(tran_data, columns=TE.columns_)

    columns = tran_df.columns

    gle = LabelEncoder()
    for i in tran_df.columns:
      tran_df[i] = gle.fit_transform(tran_df[i])
    
    tran_df = tran_df[top_items]

    df = tran_df[tran_df.sum(axis=1) >= length_trans]
    count = new_df[new_df.total_perc < sales_perc].reset_index(drop=True)

    return df, count

In [44]:
pruned_df, item_count = prune_dataset(data,2,0.4)
print(pruned_df)
#print(item_count)


      whole milk  other vegetables  ...  citrus fruit  bottled beer
1              0                 0  ...             0             0
4              1                 1  ...             0             0
5              1                 0  ...             0             0
7              0                 1  ...             0             1
10             0                 1  ...             0             0
...          ...               ...  ...           ...           ...
9829           0                 1  ...             0             0
9830           1                 0  ...             1             0
9832           0                 1  ...             1             0
9833           0                 0  ...             0             1
9834           0                 1  ...             0             0

[4585 rows x 13 columns]


In [53]:
frequent_itemsets = apriori(pruned_df, min_support=0.01, use_colnames=True)
frequent_itemsets

support                                           itemsets
0    0.442748                                       (whole milk)
1    0.352454                                 (other vegetables)
2    0.311014                                       (rolls/buns)
3    0.279389                                             (soda)
4    0.256052                                           (yogurt)
..        ...                                                ...
219  0.010687  (citrus fruit, other vegetables, tropical frui...
220  0.010033  (rolls/buns, root vegetables, yogurt, whole milk)
221  0.010469   (rolls/buns, yogurt, tropical fruit, whole milk)
222  0.012214  (root vegetables, yogurt, tropical fruit, whol...
223  0.010687  (root vegetables, yogurt, other vegetables, tr...

[224 rows x 2 columns]

In [54]:
rule_mlx = association_rules(frequent_itemsets, metric='confidence', min_threshold=0.3)
rule_mlx

antecedents  ... conviction
0                                   (other vegetables)  ...   1.023317
1                                         (whole milk)  ...   1.015857
2                                         (rolls/buns)  ...   0.914432
3                                               (soda)  ...   0.804780
4                                             (yogurt)  ...   1.050102
..                                                 ...  ...        ...
190               (yogurt, tropical fruit, whole milk)  ...   1.269141
191        (yogurt, root vegetables, other vegetables)  ...   1.303629
192          (yogurt, root vegetables, tropical fruit)  ...   1.671087
193  (root vegetables, other vegetables, tropical f...  ...   1.250245
194         (yogurt, other vegetables, tropical fruit)  ...   1.331249

[195 rows x 9 columns]

In [69]:
(rule_mlx[['antecedents','consequents','support','confidence','lift']]
.groupby('consequents')
.max()
.reset_index()
.sort_values(['lift', 'support','confidence'],
ascending=False))

consequents  ...      lift
5               (root vegetables)  ...  2.142453
8  (other vegetables, whole milk)  ...  2.076540
6                (tropical fruit)  ...  1.964469
4                        (yogurt)  ...  1.853435
1              (other vegetables)  ...  1.796927
7                       (sausage)  ...  1.720308
0                    (whole milk)  ...  1.581034
3                          (soda)  ...  1.301540
2                    (rolls/buns)  ...  1.278043

[9 rows x 5 columns]

**Conclusion:**

The itemset {yogurt, other vegetables, tropical fruit, root vegetables} is present 4.9% of the transactions and 44% of times customer who bought {yogurt, other vegetables, tropical fruit} also tend to buy root vegetables which could be affirmed by lift value of 2.14 which indicates that there is greater probability of finding root vegetables in transactions which have yogurt, other vegetables, tropical fruit than in other transactions.